File: java-breeds/app/src/main/java/es/tipolisto/breeds/data/database/RecordDao.java
package es.tipolisto.breeds.data.database;


import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

import es.tipolisto.breeds.data.model.RecordEntity;

@Dao
public interface RecordDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    Long insert(RecordEntity user);

    @Query("SELECT * FROM RecordEntity WHERE id=(:userEntityId)")
    RecordEntity getIdRecord(int userEntityId);

    @Query("SELECT * FROM RecordEntity WHERE name LIKE (:userEntityName)")
    RecordEntity getNameRecord(String userEntityName);

    @Query("SELECT * FROM RecordEntity")
    List<RecordEntity> getAllRecordEntities();

    //Obtener los últimos 10 registros ordenados por record de mayor a menor
    @Query("SELECT * FROM RecordEntity ORDER  BY score DESC LIMIT 10")
    List<RecordEntity> getLast10RecordEntities();

    @Query("SELECT * FROM RecordEntity WHERE name IN (:name)")
    List<RecordEntity> loadAllByIds(int[] name);


    @Insert
    void insertAll(√... users);

    @Update
    void update(RecordEntity user);


    @Delete
    void delete(RecordEntity user);

    //Borramos todos los que sean menores de una puntuación
    @Query("DELETE FROM RecordEntity WHERE SCORE<(:score)")
    int deleteAllMinor(int score);
}

File: kotlin-breeds/app/src/main/java/es/tipolisto/breeds/data/database/records/RecordDao.kt
package es.tipolisto.breeds.data.database.records

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update

@Dao
interface RecordDao {
    //Dame todos los records
    @Query("SELECT * FROM RecordEntity")
    fun getAll(): List<RecordEntity>

    //Dame todos los records de un usuario con un nombre determinado
    @Query("SELECT * FROM RecordEntity WHERE name LIKE (:userEntityName)")
    fun getRecordByName(userEntityName: String): RecordEntity

    //Dame el record que sea mayor que la puntuación maxima que le pasamos de los últimos 10 records
    //@Query("SELECT * FROM RecordEntity WHERE score >= (:newScore) ORDER BY score ASC LIMIT 1")

    /*
    @Query("SELECT * FROM (SELECT * FROM RecordEntity ORDER BY score DESC LIMIT 10) WHERE score >= (:newScore) ORDER BY score ASC LIMIT 1")
    fun getRecordGreaterThanScore(newScore: Int): RecordEntity?
    */

    @Query("SELECT * FROM RecordEntity WHERE score >= (:newScore) ORDER BY score ASC LIMIT 1 ")
    fun getNewRecordPosition(newScore: Int): RecordEntity?

    @Query("SELECT * FROM RecordEntity WHERE position >= (:newPosition) ORDER BY score DESC")
    fun getFollowingRecords(newPosition: Int): List<RecordEntity>?
    @Query("SELECT * FROM RecordEntity WHERE position<11 ORDER BY score DESC LIMIT 10")
    fun getFirst10RecordsByPosition(): List<RecordEntity>?
    //Obtener los últimos 10 registros ordenados por record de mayor a menor
    @Query("SELECT * FROM RecordEntity ORDER BY score DESC LIMIT 21")
    fun getFirst20RecordEntities(): List<RecordEntity>

    //Dame todos los records de un usuarios a través del id
    @Query("SELECT * FROM RecordEntity WHERE id=(:userEntityName)")
    fun getRecordById(userEntityName: Int): RecordEntity



    @Query("SELECT * FROM RecordEntity WHERE name IN (:name)")
    fun loadAllByIds(name:Array<Int>): List<RecordEntity>

    //Obtener la maxima puntación
    @Query("SELECT max(score) FROM RecordEntity")
    fun getMaxRecordEntities(): Int





    @Insert
    fun insertAll(vararg recordEntities: RecordEntity)
    @Insert
    fun insert(recordEntities: RecordEntity)

    @Update
    fun update(recordEntity: RecordEntity)

    @Delete
    fun delete(recordEntity: RecordEntity)
}